plugins/automation/app/services/discourse_automation/update.rb

```diff
@@ -0,0 +1,183 @@
+
+  def can_update_automation(guardian:)
+    guardian.is_admin?
+  end
+
+  def fetch_automation(params:)
+    DiscourseAutomation::Automation.includes(:fields, :pending_automations).find_by(
+      id: params.automation_id,
+    )
+  end
+
+  def capture_previous_state(automation:)
+    context[:previous_state] = {
+      name: automation.name,
+      script: automation.script,
+      trigger: automation.trigger,
+      enabled: automation.enabled,
+      fields: automation.serialized_fields,
+    }
+  end
+
+  def apply_forced_triggerable(automation:, params:)
+    if automation.scriptable&.forced_triggerable
+      context[:forced_trigger] = automation.scriptable.forced_triggerable[:triggerable].to_s
+    end
+  end
+
+  def handle_trigger_change(automation:, params:)
+    new_trigger = context[:forced_trigger] || params.trigger
+    return if new_trigger.blank?
+    return if automation.trigger == new_trigger
+
+    automation.fields.destroy_all
+    context[:fields_cleared] = true
+    context[:force_disable] = true
+  end
+
+  def handle_script_change(automation:, params:)
+    return if params.script.blank?
+    return if automation.script == params.script
+
+    automation.fields.destroy_all
+    context[:fields_cleared] = true
+    context[:force_disable] = true
+    context[:clear_trigger] = true
+  end
+
+  def update_automation(automation:, params:, guardian:)
+    automation.perform_required_fields_validation = true
+
+    attributes = { last_updated_by_id: guardian.user.id }
+    attributes[:name] = params.name if params.name.present?
+    attributes[:script] = params.script if params.script.present?
+    attributes[:trigger] = context[:forced_trigger] || params.trigger if params.trigger.present? ||
+      context[:forced_trigger].present?
+    attributes[:trigger] = nil if context[:clear_trigger]
+    attributes[:enabled] = params.enabled unless params.enabled.nil?
+    attributes[:enabled] = false if context[:force_disable]
+
+    # Update fields if not cleared and fields provided
+    if !context[:fields_cleared] && params.fields.present?
+      Array(params.fields)
+        .reject(&:empty?)
+        .each do |field|
+          automation.upsert_field!(
+            field[:name],
+            field[:component],
+            field[:metadata],
+            target: field[:target],
+          )
+        end
+    end
+
+    if context[:clear_trigger]
+      automation.assign_attributes(attributes)
+      automation.save!(validate: false)
+    else
+      automation.update!(attributes)
+    end
+
+    automation.reload
+  end
+
+  def log_action(automation:, guardian:)
+    previous_state = context[:previous_state]
+    changes = {}
+
+    %i[name script trigger enabled].each do |attr|
+      current_value = automation.public_send(attr)
+      previous_value = previous_state[attr]
+      if current_value != previous_value
+        changes[attr] = "#{format_value(previous_value)} → #{format_value(current_value)}"
+      end
+    end
```

Comment:
We could use `#index_with`:

```suggestion
    changes = %i[name script trigger enabled].index_with do |attr|
      current_value = automation[attr]
      previous_value = previous_state[attr]
      if current_value != previous_value
        "#{format_value(previous_value)} → #{format_value(current_value)}"
      end
    end.compact
```